### PR TITLE
Fix user authorization issues with admin sidebar

### LINKED ISSUE #1643: Users with read-only access rights should not be able to access the sidebar menu
**Describe the bug:**
An eperson group with ONLY read, default_item_read, and default_bitstream_read on items/collections/communities could still access the sidebar functions: new, edit, import, and delete.

**To reproduce**

1. Log-in as top-level admin
2. Create an eperson group (say, "regular user group") and assign add a new eperson to that group
3.  Authorize the "regular user group" to ONLY have read, default_bitstream_read, and default_item_read on items/collections/communities.
4. Log-in as that newly created eperson which belongs in the "regular user group"
5. Try accessing the sidebar menu.

The user could see: new, edit, import, and export functions. Now, these functions are rendered useless, since the user does not have any elevated roles as it only has "read" access to all items/collections/communities.

**Expected Behavior:**
This eperson group should not be able to access the sidebar, they should not be able to create new items, edit items, import metadata and export metadata. These functions should only be visible if the user has elevated roles (write, add, admin, delete, etc.)

**Related work**
Too many functions available for logged in (non-admin) persons #1483

### PR DESCRIPTION
**Pull request by KoenP**
## References
* Fixes [#1643](https://github.com/DSpace/dspace-angular/issues/1643)
* Lays groundwork for [#1331](https://github.com/DSpace/dspace-angular/issues/1331)
* [Related frontend PR](https://github.com/DSpace/dspace-angular/pull/1999)

## Description
* Frontend: hides admin sidebar sections for users who lack relevant permissions.
* Backend: adds authorization features to support the frontend changes.

## Instructions for Reviewers
In order to decide whether or not to display the "New->Item" and "Edit->Item" options in the admin sidebar, the frontend needs to know (respectively):

1. Are there any collections for which the user has permission to add new items?
2. Does the user have permission to edit at least one item?

On the backend, this patch adds two new authorization features: "canSubmit" and "canEditItem". On the frontend, the menu resolver now requests these permissions to determine whether to add the corresponding sections to the menu.

### Backend overview
Checking directly whether a user has, for instance, edit rights on at least one item is a fairly expensive operation to perform directly. The user can obtain these rights in any one of the following ways:

1. Have a direct WRITE resource policy on an item.
2. Have a direct ADMIN resource policy on an item.
3. Have ADMIN rights on a collection containing at least one item (either through a resource policy or by being a member of the collection's ADMIN group).
4. Have ADMIN rights on a community containing at least one collection that contains at least one item (either through a resource policy or by being a member of the collection's ADMIN group). Or having ADMIN rights on a community containing such a community. Or...

We don't want to hammer the DB with running the full exhaustive check every time the frontend needs this information (which is every time a user logs in), so we add this information to the index instead.

The patch introduces a new indexer plugin "SolrServiceIndexItemEditorsPlugin", and it also refactors the existing indexer plugin "SolrServiceIndexCollectionSubmittersPlugin". This means the check described above (mostly -- see Known Issues) is run on every index update instead of every time a user logs in.

#### Integration test
The patch fails `CreateMissingIdentifiersIT.testPerform` in CI but not locally. However this is known to be a flaky test and seems to have been fixed [here](https://github.com/DSpace/DSpace/pull/8597).

#### Performance
The current implementation is not as fast as it could be. For one, the two plugins are run for every item (for the canEditItem plugin) and every collection (for the canSubmit plugin), even though in both cases only a single item/collection needs to be found.

In particular, this makes recursive discovery of an ancestor community/collection for which the user has ADMIN rights less efficient. One can picture this process as a tree where the ADMIN rights "trickle down" to all descendants. The current implementation, however, computes bottom-up, starting from every collection (for the canSubmit plugin) or every item (for the canEditItem plugin), thereby potentially repeatedly exploring the same paths over and over. What's more is that, not only do we run a separate SQL query for every such item/collection involved, but every step of the recursive upward search process requires a separate DB query.

#### Known bug
The new item editing plugin was modeled on the existing collection submitters plugin, and it replicates a bug in that plugin: when recursively traversing ancestor collections/communities, it only collects the ADMIN groups that are directly registered as "the admin group" of the collections/communities, ignoring resource policies (either for groups or directly for epersons). This behavior was retained out of performance concerns with searching for research policies for every ancestor in the tree (often multiple times, as explained in the Performance subsection).

#### Proposed solutions
The known bug is easy to fix if the performance of the plugins is improved.

At the beginning of an index update, it should be possible to perform a single, relatively cheap SQL query to recursively trace out admin permissions (that is, yield a table `(group_or_eperson_id, dso_id)` and to put the result in an in-memory hash table which lives for the duration of the index update). This avoids duplicate work (because it is performed once for the entire index update, rather than for each updated object in the index) and should be relatively cheap (because in real workloads, we’d expect admin permissions to be rare compared to the number of items/collections).

Implementing these changes would require some rearchitecting; we recommend accepting this patch in the mean time.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [ ] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

**Review by tdonohue — COMMENTED** [review]
Thanks @KoenP !  This looks great overall. Code looks good and tests look good too. I also manually tested this, and it's working well.  It fixes the main bug related to the new/edit menus.  I tested these menus as a full Admin, Community Admin, Collection Admin, Submitter and a user with no rights. The menu options look correct now.

(As a sidenote, I've realized there are still other authorization issues outstanding, but those should be left for later PRs... e.g. Community Admins can _attempt_ to create a new Community anywhere, and Collection Admins can _attempt_ to edit any Collection.)

I have a few minor suggestions inline below...mostly cleanup of comments.  But, I'm basically +1 this work, as it seems to fix the reported bug.

_NOTE to other reviewers:_ You need to do a full reindex (`./dspace index-discover -b`) for these new auth checks to work properly!

File: dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java [hunk 1]
```diff
@@ -1065,6 +1075,51 @@ public boolean canEdit(Context context, Item item) throws SQLException {
...
+            Stream<String> groupIds = groupService.allMemberGroupsSet(context, currentUser).stream()
+                .map(group -> "g" + group.getID());
+            String query = Stream.concat(Stream.of(userId), groupIds)
+                .collect(Collectors.joining(" OR ", "edit:(", ")"));
+            discoverQuery.addFilterQueries(query);
+        }
+        return searchService.search(context, discoverQuery);
+    }
+
+    public List<Item> findItemsWithEdit(Context context, int offset, int limit)
```
[the remark this answers is not in the document]
Comment (KoenP) [reply dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java:1104]: Hi @tdonohue, thanks for your feedback! I added the annotation.

File: dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java [hunk 2]
```diff
@@ -1065,6 +1075,51 @@ public boolean canEdit(Context context, Item item) throws SQLException {
...
+        discoverQuery.setDSpaceObjectFilter(IndexableItem.TYPE);
+        discoverQuery.setStart(offset);
+        discoverQuery.setMaxResults(limit);
+        DiscoverResult resp = retrieveItemsWithEdit(context, discoverQuery);
+        return resp.getIndexableObjects().stream()
+            .map(solrItems -> ((IndexableItem) solrItems).getIndexedObject())
+            .collect(Collectors.toList());
+    }
+
+    public int countItemsWithEdit(Context context) throws SQLException, SearchServiceException {
```
[the remark this answers is not in the document]
Comment (KoenP) [reply dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java:1117]: I added the annotation.

File: dspace-api/src/main/java/org/dspace/discovery/IndexingUtils.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+public class IndexingUtils {
+    private IndexingUtils() {
+    }
+
+    /**
+     * Retrieve all ancestor communities of a given community, with the first one being the given community and the
+     * last one being the root.
+     * <p>
+     * TODO: can be done in a single SQL query with recursive common table expressions
+     * TODO: should probably be moved to CommunityService
```
[the remark this answers is not in the document]
Comment (KoenP) [reply dspace-api/src/main/java/org/dspace/discovery/IndexingUtils.java]: I removed these TODOs as I don't plan on adding them to the PR and currently there doesn't seem to be a pressing reason to act on them.

File: dspace-api/src/main/java/org/dspace/discovery/SolrServiceIndexCollectionSubmittersPlugin.java [hunk 4]
```diff
@@ -42,29 +43,21 @@ public void additionalIndex(Context context, IndexableObject idxObj, SolrInputDo
...
-                        if (parent.getAdministrators() != null) {
-                            fieldValue = "g" + parent.getAdministrators().getID();
-                            document.addField("submit", fieldValue);
-                        }
-                        parent = (Community) ContentServiceFactory.getInstance().getDSpaceObjectService(parent)
-                                                                                .getParentObject(context, parent);
+                    // Index groups with ADMIN rights on the Collection, on
+                    // Communities containing those Collections, and recursively on any Community containing such a
+                    // Community.
+                    // TODO: Strictly speaking we should also check for epersons who received admin rights directly,
```
[the remark this answers is not in the document]
Comment (KoenP) [reply dspace-api/src/main/java/org/dspace/discovery/SolrServiceIndexCollectionSubmittersPlugin.java:49]: This TODO is still relevant and I don't think we should remove it. It is related to the "Known bug" section of the PR.

What the method does now is:

- Collect _admin groups_ for the collection and for communities (transitively) containing the currently considered collection.
- Collect epersons and groups with admin or edit rights _by resource policy_ on the collection.

However, what we currently don't do is collecting epersons or groups with _admin rights by resource policy_ on communities (transitively) containing the collection.

File: dspace-api/src/main/java/org/dspace/discovery/SolrServiceIndexItemEditorsPlugin.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    @Override
+    public void additionalIndex(Context context, IndexableObject idxObj, SolrInputDocument document) {
+        if (idxObj instanceof IndexableItem) {
+            Item item = ((IndexableItem) idxObj).getIndexedObject();
+            if (item != null) {
+                try {
+                    // Index groups with ADMIN rights on Collections containing the Item, on
+                    // Communities containing those Collections, and recursively on any Community containing ssuch a
+                    // Community.
+                    // TODO: Strictly speaking we should also check for epersons who received admin rights directly,
```
[the remark this answers is not in the document]
Comment (KoenP) [reply dspace-api/src/main/java/org/dspace/discovery/SolrServiceIndexItemEditorsPlugin.java:49]: See above response.

File: dspace-api/src/main/java/org/dspace/content/service/ItemService.java [hunk 6]
```diff
@@ -768,6 +769,27 @@ public Iterator<Item> findByLastModifiedSince(Context context, Date last)
...
+    /**
+     * finds all items for which the current user has editing rights
+     * @param context DSpace context object
+     * @param offset page offset
+     * @param limit  page size limit
+     * @return list of items for which the current user has editing rights
+     * @throws SQLException
+     * @throws SearchServiceException
+     */
+    public List<Item> findItemsWithEdit(Context context, int offset, int limit)
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/service/ItemService.java:781]: Could we add Tests for these new methods to the existing `org.dspace.content.ItemTest` class?  I just realized we're missing tests for `findItemsWithEdit()` and `countItemsWithEdit()`.

There are similar tests in that class for other `find*()` methods and even `canEdit()` for the ItemService. So, it'd be good to add a few small tests for these new methods.

**Review by tdonohue — COMMENTED** [review]
@KoenP : Thanks for the updates!  The updates & your answers (to my other questions) all make sense.  I gave this another look today, and realized that I missed a small detail in my earlier review.  It looks like we're accidentally missing tests for the new ItemService methods.  If those could be added, I'd appreciate it.

Again, this doesn't change the functionality. This feature works, so I'm basically a +1.  But, tests are required for new methods (unless there's a reason they are overly complex to create).  Thanks!

File: dspace-api/src/main/java/org/dspace/content/service/ItemService.java (hunk 6 above)
Comment (KoenP) [reply dspace-api/src/main/java/org/dspace/content/service/ItemService.java:781]: Hi @tdonohue, I added some tests to the `ItemServiceTest` class, as apparently there's no SOLR indexing available in the `ItemTest` tests. I kept the tests rather simple, as the full functionality is tested extensively in `EditItemFeatureIT`.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @KoenP for the additional tests!  This now looks good to me & works as described.

@atarix83 : If you can find a chance to test this alongside dspace/dspace-angular#1999, I'd appreciate it.  I think these two PRs are ready to merge, and they are required for additional permission fixes to the sidebar menu.

**Comment by tdonohue** [comment]
@atarix83 or @LucaGiamminonni : Are you still planning to give a quick review of this (and the corresponding https://github.com/DSpace/dspace-angular/pull/1999) ?

These PRs are required for others to also be merged (namely #8616 and https://github.com/DSpace/dspace-angular/pull/2011)... and we are starting to build a backlog of PRs that depend on these.  If you could get to these reviews soon (in the next few days), I'd appreciate it.

File: dspace-api/src/main/java/org/dspace/discovery/SolrServiceIndexCollectionSubmittersPlugin.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/discovery/SolrServiceIndexCollectionSubmittersPlugin.java:49]: is there a reason why this gap is not fixed right now? I mean the solution for the collection can easily be replicated for the community, right?

File: dspace/solr/search/conf/schema.xml [hunk 7]
```diff
@@ -264,6 +264,9 @@
     <!-- used to track which group(s) have submit permissions -->
     <field name="submit" type="string" indexed="true" stored="true" omitNorms="true" multiValued="true" docValues="true" />
 
+    <!-- used to track which eperson(s) have edit permissions on items -->
```
Comment (abollini) [inline dspace/solr/search/conf/schema.xml:267]: it should be eperson(s) and/or group(s)

**Review by abollini — APPROVED** [review]
thanks @KoenP for this PR, it looks good to me. Just two minor comments inline

**Comment by tdonohue** [comment]
Merging as this PR & the corresponding frontend are both now at +2.  Thanks again @KoenP !

**Pull request merged**